### PR TITLE
Generate unique instance ID for each cloud..

### LINKED ISSUE #46: One-shot feature kills slaves of unrelated projects
Greetings,

We've installed v2.0 on several of our jenkins masters, and encountered the following issue:

Our jenkins masters are independant instances, most teams run their own masters. But slaves are launched and run in one google project, maintained by IT. For some masters (teams), new "one-shot" feature is enabled, for others, it is not. And some remaining masters are still running older versions of the plugin without these new one-shot and snapshot features at all.

Apparently, some of the masters with one-shot enabled will start killing slaves of other masters, where one-shot is disabled, probably following some naming pattern which can be similar for different masters. As a result, some teams end up with their slaves killed by other teams' masters while jobs are still run: e.g. "Team A" slave gets killed by "Team B" master, which are two independent teams.

### PR DESCRIPTION
**Pull request by ingwarsw**
Generate unique ID UUID for each cloud.

- Resolves issues with #36 
- Closes #46 
- Closes https://issues.jenkins-ci.org/browse/JENKINS-56201

### CODE REVIEW

File: src/test/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloudTest.java [hunk 1]
```diff
@@ -92,6 +94,13 @@ public void construction() throws Exception {
             Assert.assertEquals("Cloud reference was not set on child InstanceConfiguration", ic.cloud, cloud);
         }
     }
+    
+    @Test
+    public void instanceIdWasGenerated() throws Exception {
+        ComputeEngineCloud cloud = new ComputeEngineCloud(INSTANCE_ID, CLOUD_NAME, PROJECT_ID, PROJECT_ID, INSTANCE_CAP_STR, null);
+
```
Comment (craigatgoogle) [inline src/test/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloudTest.java]: Remove extra '\n'.

File: src/test/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloudTest.java [hunk 2]
```diff
@@ -92,6 +94,13 @@ public void construction() throws Exception {
...
+        Assert.assertNotNull("Instance ID was not generated in constructor", cloud.getInstanceId());
```
Comment (craigatgoogle) [inline src/test/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloudTest.java]: Suggest also testing to ensure the string is not empty.

**Comment by craigatgoogle** [comment]
@rachely3n getInstanceUniqueId() returned the hashcode of the cloud's name, which would result in a collision if two clouds shared the same name. This PR generates a unique UUID for each cloud and incorporates it into the cloud label, which removes potential collisions.

**Comment by rachely3n** [comment]
@craigatgoogle Yep, I understood that portion, but just wanted to verify, is the problem with two different masters only defined by different clouds or do we also want to distinguish instances under different instance configurations.

**Comment by craigatgoogle** [comment]
@rachely3n the referenced issue is #46 which is two masters using the same cloud configuration. This PR would solve that problem.

**Comment by ingwarsw** [comment]
> Small comment; everything else looks good. However, I'm not sure I see how this change will close out those issues. Instead of getInstanceUniqueId, we're replacing it with UUID. Wouldn't instances under different configurations but with the same cloud instance id still be identified as the same by CleanLostNodes?

No..
That never happened...
Both issues (#36 and #46) was about same names (ids) of different clouds.

**Comment by ingwarsw** [comment]
@rachely3n Maybe in some ways..
Only think that might be wort adding at instance configuration level is max number of instances..
But if really needed it can be now easily done by creating 2 or more clouds..

**Comment by rachely3n** [comment]
Just ran the integration tests. One test is failing:
``` ComputeEngineCloudIT.testTemplate:376 Mar 22, 2019 10:58:19 AM com.google.jenkins.plugins.computeengine.ComputeEngineCloud provision
INFO: Provisioning node from config com.google.jenkins.plugins.computeengine.InstanceConfiguration@475e0463 for excess workload of 1 units of label 'integration'
Mar 22, 2019 10:58:19 AM com.google.jenkins.plugins.computeengine.ComputeEngineCloud availableNodeCapacity
INFO: Found capacity for 10 nodes in cloud integration
Mar 22, 2019 10:58:21 AM com.google.jenkins.plugins.computeengine.ComputeEngineCloud lambda$provision$0
INFO: Waiting 300000ms for node integration-qn1zfz to connect
Mar 22, 2019 10:58:22 AM com.google.jenkins.plugins.computeengine.ComputeEngineCloud provision
INFO: Provisioning node from config com.google.jenkins.plugins.computeengine.InstanceConfiguration@485cb497 for excess workload of 1 units of label 'integration'
Mar 22, 2019 10:58:23 AM com.google.jenkins.plugins.computeengine.ComputeEngineCloud availableNodeCapacity
INFO: Found capacity for 9 nodes in cloud integration
Mar 22, 2019 10:58:25 AM com.google.jenkins.plugins.computeengine.ComputeEngineCloud lambda$provision$0
INFO: Waiting 300000ms for node integration-pmrzns to connect
Mar 22, 2019 10:59:55 AM com.google.jenkins.plugins.computeengine.ComputeEngineCloud lambda$provision$0
INFO: 90796ms elapsed waiting for node integration-pmrzns to connect
 expected:<[954519824]> but was:<[81926d94-678f-4607-a49c-d519f87b5faf]>
[INFO] 
[ERROR] Tests run: 12, Failures: 1, Errors: 0, Skipped: 0
```

**Comment by ingwarsw** [comment]
@rachely3n Which test is failing?
I cannot see `954519824` as static anywhere..

**Comment by ingwarsw** [comment]
@rachely3n Ok found it :)

**Comment by ingwarsw** [comment]
Is there a way we could run integration tests in CI?
Maybe only from inside local (from inside repo itself) PRs but still would be great..

**Comment by rachely3n** [comment]
@ingwarsw we're currently trying to improve that part, but we haven't been able to devote much resources to it. Sorry about that! I know it can be frustrating.

**Comment by craigatgoogle** [comment]
Thanks again for contributing this, super useful!
